### PR TITLE
Prompt the user to confirm if similar redirects already exist

### PR DESCRIPTION
**Pull request by kalleth**
# [Trello card](https://trello.com/c/KVfwkSPW/18-indicate-pre-existing-from-and-target-urls-in-short-url-requests)

## Screenshot
screenshot from 2016-09-05 15-15-40

## Rationale
To reduce the volume of duplicates appearing in the short URL manager, we can show the user a confirmation message with information about live short URLs and ask them if they really, really want to proceed!

## Included refactor
5219869 is worth reviewing in isolation; it doesn't change the included logic but does move it to command objects. The actual functionality in this PR is added in b2bf138.

### CODE REVIEW

**Comment by elliotcm** [comment]
Broadly looks good.  Would like @NeilvB to also look over it since he knows the app better than me.

File: app/controllers/short_url_requests_controller.rb [hunk 1]
```diff
@@ -24,12 +24,16 @@ def create
     @short_url_request.requester = current_user
     @short_url_request.contact_email = current_user.email
 
-    if @short_url_request.save
-      Notifier.short_url_requested(@short_url_request).deliver_now
-      flash[:success] = "Your request has been made."
-      redirect_to root_path
+    if @short_url_request.duplicate? && !@short_url_request.confirmed
```
Comment (NeilvB) [inline app/controllers/short_url_requests_controller.rb]: We could move these checks into a single method on the model perhaps: `@short_url_request.requires_confirmation` ?

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -7,4 +7,8 @@ def crumb_li(title, path, active)
       active ? title : link_to(title, path)
     end
   end
+
+  def overwrites_existing?(short_url_request)
```
Comment (NeilvB) [inline app/helpers/application_helper.rb]: I don't feel too strongly but I wonder if `would_overwrite_existing?` might be a small readability improvement for the method name here. My thinking is that at the point when this is called, the request isn't confirmed so it doesn't overwrite an existing request.

File: app/models/short_url_request.rb [hunk 3]
```diff
@@ -25,6 +25,16 @@ class ShortUrlRequest
 
   scope :pending, -> { where(state: "pending") }
 
+  attr_accessor :confirmed
+
+  def duplicates
+    @duplicates ||= Redirect.or({from_path: from_path}, {to_path: to_path})
```
Comment (NeilvB) [inline app/models/short_url_request.rb]: I wonder if we want to be more strict about what our definition of a "duplicate" request is. We have a story in progress which prevents "full duplicates"  - as in, requests that match the from path and to path for an existing live redirect. Is there therefore a case for naming this method to something else to make this distinction clear, like `redirects_with_matching_paths`?

File: app/controllers/short_url_requests_controller.rb (hunk 1 above)
Comment (kalleth) [inline app/controllers/short_url_requests_controller.rb]: I'll play with this -- that's a possibility.

File: app/helpers/application_helper.rb (hunk 2 above)
Comment (kalleth) [inline app/helpers/application_helper.rb]: I think you're right. I will change the method.

File: app/models/short_url_request.rb (hunk 3 above)
Comment (kalleth) [inline app/models/short_url_request.rb]: How about `similar_redirects` ? I was planning on adding the "full duplicates" story as a simple validator method separate to these.

File: spec/controllers/short_url_requests_controller_spec.rb [hunk 4]
```diff
@@ -139,20 +135,53 @@
...
+              confirmed: true,
+            }
+          } }
+
+          it "creates a short url request" do
+            post :create, params
+            expect(ShortUrlRequest.count).to eq(1)
+          end
+        end
+      end
```
Comment (NeilvB) [inline spec/controllers/short_url_requests_controller_spec.rb]: nice specs :)

File: spec/lib/commands/short_url_requests/accept_spec.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  context "with a successful redirect creation" do
+    it "results in an active redirect" do
+      command.call(failure: failure)
+
+      expect(url_request.redirect).not_to be_nil
+      expect(url_request.redirect.attributes.slice(:from_path, :to_path)).to eq(url_request.attributes.slice(:from_path, :to_path))
+    end
+
+    it "changes state to accepted" do
```
Comment (NeilvB) [inline spec/lib/commands/short_url_requests/accept_spec.rb]: nit: changes request state to accepted?

File: lib/commands/short_url_requests/create.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  def create_object(params, requester)
+    ShortUrlRequest.new(params).tap do |url_request|
+      url_request.requester = requester
+      url_request.contact_email = requester.email
+    end
+  end
+
+  def requires_confirmation?(url_request)
+    url_request.duplicate? && !url_request.confirmed
```
Comment (NeilvB) [inline lib/commands/short_url_requests/create.rb]: 👍

File: app/models/short_url_request.rb (hunk 3 above)
Comment (NeilvB) [inline app/models/short_url_request.rb]: Like the sound of that ! 👍

**Comment by NeilvB** [comment]
Overall this looks great. I made a few minor comments.

Moving the "management" logic that's concerned with keeping the requests and redirects in the correct states into separate classes feels like a good step. I wasn't super happy with having the Request class know all about Redirects. We're deviating from ultra-standard Rails controller methods here, but I feel like it's a good tradeoff.

**Comment by kalleth** [comment]
Got a final 👍 from @NeilvB so i'm merging.

**Pull request merged**